# Advanced MLP - 2
- Putting it altogether

In [1]:
import matplotlib.pyplot as plt
import numpy as np

from sklearn.ensemble import VotingClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from keras.datasets import mnist
from keras.wrappers.scikit_learn import KerasClassifier
from keras.datasets import mnist
from keras.models import Sequential
from keras.utils.np_utils import to_categorical
from keras.models import Sequential
from keras.layers import Activation, Dense, BatchNormalization, Dropout
from keras import optimizers

Using TensorFlow backend.


## Load Dataset

In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [3]:
# reshaping X data: (n, 28, 28) => (n, 784)
X_train = X_train.reshape((X_train.shape[0], X_train.shape[1] * X_train.shape[2]))
X_test = X_test.reshape((X_test.shape[0], X_test.shape[1] * X_test.shape[2]))

In [4]:
# We use all training data and validate on all test data
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(60000, 784)
(10000, 784)
(60000,)
(10000,)


## Training & Validating Model
- Measures to improve training is applied simultaneously
    - More training set
    - Weight Initialization scheme
    - Nonlinearity (Activation function)
    - Optimizers: adaptvie
    - Batch Normalization
    - Dropout (Regularization)
    - Model Ensemble

In [5]:
def mlp_model():
    model = Sequential()
    
    model.add(Dense(50, input_shape = (784, ), kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(Dropout(0.2))
    model.add(Dense(50, kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))    
    model.add(Dropout(0.2))
    model.add(Dense(50, kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(Dropout(0.2))
    model.add(Dense(50, kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(Dropout(0.2))
    model.add(Dense(10, kernel_initializer='he_normal'))
    model.add(Activation('softmax'))
    
    adam = optimizers.Adam(lr = 0.001)
    model.compile(optimizer = adam, loss = 'categorical_crossentropy', metrics = ['accuracy'])
    
    return model

In [6]:
# create 5 models to ensemble
model1 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model2 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model3 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model4 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model5 = KerasClassifier(build_fn = mlp_model, epochs = 100)

In [7]:
ensemble_clf = VotingClassifier(estimators = [('model1', model1), ('model2', model2), ('model3', model3), ('model4', model4), ('model5', model5)], voting = 'soft')

In [8]:
ensemble_clf.fit(X_train, y_train)

Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
Instructions for updating:
keep_dims is deprecated, use keepdims instead
Instructions for updating:
Use tf.cast instead.
Epoch 1/100
60000/60000 [==============================] - 8s - loss: 0.7211 - acc: 0.7824     
Epoch 2/100
60000/60000 [==============================] - 7s - loss: 0.4177 - acc: 0.8799     
Epoch 3/100
60000/60000 [==============================] - 6s - loss: 0.3612 - acc: 0.8983     
Epoch 4/100
60000/60000 [==============================] - 7s - loss: 0.3179 - acc: 0.9107     
Epoch 5/100
60000/60000 [==============================] - 7s - loss: 0.2895 - acc: 0.9188     
Epoch 6/100
60000/60000 [==============================] - 7s - loss: 0.2759 - acc: 0.9229     
Epoch 7/100
60000/60000 [==============================] - 7s - loss: 0.2645 - acc: 0.9244     
Epoch 8/100
60000/60

60000/60000 [==============================] - 11s - loss: 0.1615 - acc: 0.9531    
Epoch 58/100
60000/60000 [==============================] - 12s - loss: 0.1575 - acc: 0.9544    
Epoch 59/100
60000/60000 [==============================] - 12s - loss: 0.1610 - acc: 0.9527    
Epoch 60/100
60000/60000 [==============================] - 12s - loss: 0.1578 - acc: 0.9542    
Epoch 61/100
60000/60000 [==============================] - 12s - loss: 0.1592 - acc: 0.9536    
Epoch 62/100
60000/60000 [==============================] - 11s - loss: 0.1565 - acc: 0.9546    
Epoch 63/100
60000/60000 [==============================] - 10s - loss: 0.1562 - acc: 0.9556    
Epoch 64/100
60000/60000 [==============================] - 10s - loss: 0.1558 - acc: 0.9543    
Epoch 65/100
60000/60000 [==============================] - 11s - loss: 0.1563 - acc: 0.9548    
Epoch 66/100
60000/60000 [==============================] - 10s - loss: 0.1516 - acc: 0.9557    
Epoch 67/100
60000/60000 [=================

60000/60000 [==============================] - 6s - loss: 0.1904 - acc: 0.9442     
Epoch 25/100
60000/60000 [==============================] - 6s - loss: 0.1890 - acc: 0.9456     
Epoch 26/100
60000/60000 [==============================] - 6s - loss: 0.1863 - acc: 0.9463     
Epoch 27/100
60000/60000 [==============================] - 7s - loss: 0.1912 - acc: 0.9447     
Epoch 28/100
60000/60000 [==============================] - 6s - loss: 0.1877 - acc: 0.9456     
Epoch 29/100
60000/60000 [==============================] - 6s - loss: 0.1827 - acc: 0.9460     
Epoch 30/100
60000/60000 [==============================] - 6s - loss: 0.1826 - acc: 0.9468     
Epoch 31/100
60000/60000 [==============================] - 7s - loss: 0.1832 - acc: 0.9468     
Epoch 32/100
60000/60000 [==============================] - 7s - loss: 0.1846 - acc: 0.9468     
Epoch 33/100
60000/60000 [==============================] - 7s - loss: 0.1781 - acc: 0.9483     
Epoch 34/100
60000/60000 [=================

60000/60000 [==============================] - 7s - loss: 0.1440 - acc: 0.9585     
Epoch 93/100
60000/60000 [==============================] - 7s - loss: 0.1448 - acc: 0.9590     
Epoch 94/100
60000/60000 [==============================] - 7s - loss: 0.1425 - acc: 0.9593     
Epoch 95/100
60000/60000 [==============================] - 7s - loss: 0.1394 - acc: 0.9588     
Epoch 96/100
60000/60000 [==============================] - 7s - loss: 0.1457 - acc: 0.9582     
Epoch 97/100
60000/60000 [==============================] - 6s - loss: 0.1415 - acc: 0.9582     
Epoch 98/100
60000/60000 [==============================] - 7s - loss: 0.1403 - acc: 0.9586     
Epoch 99/100
60000/60000 [==============================] - 7s - loss: 0.1410 - acc: 0.9587     
Epoch 100/100
60000/60000 [==============================] - 7s - loss: 0.1379 - acc: 0.9603     


VotingClassifier(estimators=[('model1',
                              <keras.wrappers.scikit_learn.KerasClassifier object at 0x131edfb38>),
                             ('model2',
                              <keras.wrappers.scikit_learn.KerasClassifier object at 0x131edfb70>),
                             ('model3',
                              <keras.wrappers.scikit_learn.KerasClassifier object at 0x131edfa90>),
                             ('model4',
                              <keras.wrappers.scikit_learn.KerasClassifier object at 0x131edfac8>),
                             ('model5',
                              <keras.wrappers.scikit_learn.KerasClassifier object at 0x131edfbe0>)],
                 flatten_transform=True, n_jobs=None, voting='soft',
                 weights=None)

In [9]:
y_pred = ensemble_clf.predict(X_test)

 9984/10000 [============================>.] - ETA: 0s

In [10]:
print('Acc: ', accuracy_score(y_pred, y_test))

Acc:  0.9794


Final model produces near **98% of accuracy**, which is rather improved from before